File: tests/test_command_context.py
# pyright: standard

import json
from pathlib import Path

from typer.testing import CliRunner

from aico.lib.session_find import complete_files_in_context
from aico.main import app

runner = CliRunner()


def test_add_file_to_context(tmp_path: Path) -> None:
    # GIVEN an initialized session and a file to add
    with runner.isolated_filesystem(temp_dir=tmp_path) as td:
        runner.invoke(app, ["init"])

        test_file = Path(td) / "test_file.py"
        test_file.write_text("print('hello')")

        # WHEN `aico add` is run with the file path
        result = runner.invoke(app, ["add", "test_file.py"])

        # THEN the command succeeds and reports the addition
        assert result.exit_code == 0
        assert "Added file to context: test_file.py" in result.stdout

        # AND the session file is updated with the file's relative path
        view_file = Path(td) / ".aico" / "sessions" / "main.json"
        view_data = json.loads(view_file.read_text())
        assert view_data["context_files"] == ["test_file.py"]


def test_add_duplicate_file_is_ignored(tmp_path: Path) -> None:
    # GIVEN a session with a file already in the context
    with runner.isolated_filesystem(temp_dir=tmp_path) as td:
        runner.invoke(app, ["init"])
        test_file = Path(td) / "test_file.py"
        test_file.write_text("print('hello')")
        # Add it once
        runner.invoke(app, ["add", str(test_file)])

        # WHEN the same file is added again
        result = runner.invoke(app, ["add", str(test_file)])

        # THEN the command reports that the file is already in context
        assert result.exit_code == 0
        assert "File already in context: test_file.py" in result.stdout

        # AND the session context list remains unchanged
        view_file = Path(td) / ".aico" / "sessions" / "main.json"
        view_data = json.loads(view_file.read_text())
        assert view_data["context_files"] == ["test_file.py"]


def test_add_non_existent_file_fails(tmp_path: Path) -> None:
    # GIVEN an initialized session
    with runner.isolated_filesystem(temp_dir=tmp_path):
        runner.invoke(app, ["init"])

        # WHEN adding a file that does not exist
        result = runner.invoke(app, ["add", "non_existent_file.py"])

        # THEN the command fails with an error
        assert result.exit_code == 1
        assert "Error: File not found: non_existent_file.py" in result.stderr


def test_add_file_outside_session_root_fails(tmp_path: Path) -> None:
    # GIVEN a session in one directory and a file in a parallel directory
    project_dir = tmp_path / "project"
    project_dir.mkdir()

    other_dir = tmp_path / "other"
    other_dir.mkdir()

    other_file = other_dir / "file.txt"
    other_file.touch()

    with runner.isolated_filesystem(temp_dir=project_dir) as td:
        runner.invoke(app, ["init"])

        # WHEN attempting to add the file using a path that goes outside the session root
        # Note: We resolve the path to be absolute to test the logic robustly.
        result = runner.invoke(app, ["add", str(other_file.resolve())])

        # THEN the command fails with a clear error message
        assert result.exit_code == 1
        assert (
            f"Error: File '{other_file.resolve()}' is outside the session root '{Path(td).resolve()}'" in result.stderr
        )


def test_add_multiple_files_successfully(tmp_path: Path) -> None:
    # GIVEN an initialized session and two files to add
    with runner.isolated_filesystem(temp_dir=tmp_path) as td:
        runner.invoke(app, ["init"])

        file1 = Path(td) / "file1.py"
        file1.write_text("content1")
        file2 = Path(td) / "file2.py"
        file2.write_text("content2")

        # WHEN `aico add` is run with multiple files
        result = runner.invoke(app, ["add", "file1.py", "file2.py"])

        # THEN the command succeeds and reports both additions
        assert result.exit_code == 0
        assert "Added file to context: file1.py" in result.stdout
        assert "Added file to context: file2.py" in result.stdout

        # AND the session file is updated with both relative paths
        view_file = Path(td) / ".aico" / "sessions" / "main.json"
        view_data = json.loads(view_file.read_text())
        assert sorted(view_data["context_files"]) == ["file1.py", "file2.py"]


def test_add_multiple_files_with_one_already_in_context(tmp_path: Path) -> None:
    # GIVEN a session with one file already in context
    with runner.isolated_filesystem(temp_dir=tmp_path) as td:
        runner.invoke(app, ["init"])

        file1 = Path(td) / "file1.py"
        file1.write_text("content1")
        file2 = Path(td) / "file2.py"
        file2.write_text("content2")
        runner.invoke(app, ["add", "file1.py"])  # Pre-add file1

        # WHEN `aico add` is run with both the existing and a new file
        result = runner.invoke(app, ["add", "file1.py", "file2.py"])

        # THEN the command succeeds and reports the correct status for each
        assert result.exit_code == 0
        assert "File already in context: file1.py" in result.stdout
        assert "Added file to context: file2.py" in result.stdout

        # AND the session file contains both files without duplicates
        view_file = Path(td) / ".aico" / "sessions" / "main.json"
        view_data = json.loads(view_file.read_text())
        assert sorted(view_data["context_files"]) == ["file1.py", "file2.py"]


def test_add_multiple_files_with_one_non_existent_partially_fails(
    tmp_path: Path,
) -> None:
    # GIVEN an initialized session and one valid and one non-existent file
    with runner.isolated_filesystem(temp_dir=tmp_path) as td:
        runner.invoke(app, ["init"])

        file1 = Path(td) / "file1.py"
        file1.write_text("content1")
        non_existent_file = "non_existent.py"

        # WHEN `aico add` is run with both files
        result = runner.invoke(app, ["add", "file1.py", non_existent_file])

        # THEN the command exits with a non-zero status code
        assert result.exit_code == 1

        # AND it reports the success for the valid file
        assert "Added file to context: file1.py" in result.stdout

        # AND it reports an error for the non-existent file
        assert f"Error: File not found: {non_existent_file}" in result.stderr

        # AND the session file is updated with only the valid file
        view_file = Path(td) / ".aico" / "sessions" / "main.json"
        view_data = json.loads(view_file.read_text())
        assert view_data["context_files"] == ["file1.py"]


def test_drop_single_file_successfully(tmp_path: Path) -> None:
    # GIVEN a session with two files in context
    with runner.isolated_filesystem(temp_dir=tmp_path) as td:
        runner.invoke(app, ["init"])
        (Path(td) / "file1.py").touch()
        (Path(td) / "file2.py").touch()
        runner.invoke(app, ["add", "file1.py", "file2.py"])

        # WHEN `aico drop` is run on one file
        result = runner.invoke(app, ["drop", "file1.py"])

        # THEN the command succeeds and reports the removal
        assert result.exit_code == 0
        assert "Dropped file from context: file1.py" in result.stdout

        # AND the session file is updated to contain only the other file
        view_file = Path(td) / ".aico" / "sessions" / "main.json"
        view_data = json.loads(view_file.read_text())
        assert sorted(view_data["context_files"]) == ["file2.py"]


def test_drop_multiple_files_successfully(tmp_path: Path) -> None:
    # GIVEN a session with three files in context
    with runner.isolated_filesystem(temp_dir=tmp_path) as td:
        runner.invoke(app, ["init"])
        (Path(td) / "file1.py").touch()
        (Path(td) / "file2.py").touch()
        (Path(td) / "file3.py").touch()
        runner.invoke(app, ["add", "file1.py", "file2.py", "file3.py"])

        # WHEN `aico drop` is run on two files
        result = runner.invoke(app, ["drop", "file1.py", "file3.py"])

        # THEN the command succeeds and reports both removals
        assert result.exit_code == 0
        assert "Dropped file from context: file1.py" in result.stdout
        assert "Dropped file from context: file3.py" in result.stdout

        # AND the session file is updated correctly
        view_file = Path(td) / ".aico" / "sessions" / "main.json"
        view_data = json.loads(view_file.read_text())
        assert sorted(view_data["context_files"]) == ["file2.py"]


def test_drop_file_not_in_context_fails(tmp_path: Path) -> None:
    # GIVEN a session with one file in context
    with runner.isolated_filesystem(temp_dir=tmp_path) as td:
        runner.invoke(app, ["init"])
        (Path(td) / "file1.py").touch()
        runner.invoke(app, ["add", "file1.py"])

        # WHEN `aico drop` is run on a file not in the context
        result = runner.invoke(app, ["drop", "not_in_context.py"])

        # THEN the command fails with a non-zero exit code
        assert result.exit_code == 1

        # AND an error is printed to stderr
        assert "Error: File not in context: not_in_context.py" in result.stderr

        # AND the session file remains unchanged
        view_file = Path(td) / ".aico" / "sessions" / "main.json"
        view_data = json.loads(view_file.read_text())
        assert view_data["context_files"] == ["file1.py"]


def test_drop_multiple_with_one_not_in_context_partially_fails(tmp_path: Path) -> None:
    # GIVEN a session with two files in context
    with runner.isolated_filesystem(temp_dir=tmp_path) as td:
        runner.invoke(app, ["init"])
        (Path(td) / "file1.py").touch()
        (Path(td) / "file2.py").touch()
        runner.invoke(app, ["add", "file1.py", "file2.py"])

        # WHEN `aico drop` is run with one valid and one invalid file
        result = runner.invoke(app, ["drop", "file1.py", "not_in_context.py"])

        # THEN the command fails with a non-zero exit code
        assert result.exit_code == 1

        # AND it reports the successful removal
        assert "Dropped file from context: file1.py" in result.stdout

        # AND it reports the error for the other file
        assert "Error: File not in context: not_in_context.py" in result.stderr

        # AND the session file is updated to remove the valid file
        view_file = Path(td) / ".aico" / "sessions" / "main.json"
        view_data = json.loads(view_file.read_text())
        assert sorted(view_data["context_files"]) == ["file2.py"]


def test_drop_autocompletion(tmp_path: Path) -> None:
    # GIVEN a session with several files in context
    with runner.isolated_filesystem(temp_dir=tmp_path) as td:
        # AND a session file is initialized with context files
        result = runner.invoke(app, ["init"])
        assert result.exit_code == 0
        view_file = Path(td) / ".aico" / "sessions" / "main.json"
        view_data = json.loads(view_file.read_text())
        view_data["context_files"] = [
            "src/main.py",
            "src/utils.py",
            "docs/README.md",
        ]
        view_file.write_text(json.dumps(view_data))

        # WHEN the completion function is called with various partial inputs
        # THEN it returns the correct list of matching files
        assert sorted(complete_files_in_context(None, [], "src/")) == [
            "src/main.py",
            "src/utils.py",
        ]
        assert complete_files_in_context(None, [], "docs/") == ["docs/README.md"]
        assert complete_files_in_context(None, [], "src/main") == ["src/main.py"]
        assert complete_files_in_context(None, [], "invalid") == []


def test_add_symlink_to_inside_success(tmp_path: Path) -> None:
    # GIVEN an initialized session and a target file
    with runner.isolated_filesystem(temp_dir=tmp_path) as td:
        runner.invoke(app, ["init"])

        target = Path(td) / "target.py"
        target.write_text("print('hello')")

        sym = Path(td) / "sym.py"
        sym.symlink_to(target)

        # WHEN adding the symlink
        result = runner.invoke(app, ["add", "sym.py"])

        # THEN succeeds and adds the symlink path
        assert result.exit_code == 0
        assert "Added file to context: sym.py" in result.stdout

        # AND session contains the symlink relative path
        view_file = Path(td) / ".aico" / "sessions" / "main.json"
        view_data = json.loads(view_file.read_text())
        assert view_data["context_files"] == ["sym.py"]


def test_drop_symlink_success(tmp_path: Path) -> None:
    # GIVEN session with symlink in context
    with runner.isolated_filesystem(temp_dir=tmp_path) as td:
        runner.invoke(app, ["init"])

        target = Path(td) / "target.py"
        target.write_text("content")
        sym = Path(td) / "sym.py"
        sym.symlink_to(target)
        runner.invoke(app, ["add", "sym.py"])

        # WHEN dropping the symlink path
        result = runner.invoke(app, ["drop", "sym.py"])

        # THEN succeeds
        assert result.exit_code == 0
        assert "Dropped file from context: sym.py" in result.stdout

        # AND removed from session
        view_file = Path(td) / ".aico" / "sessions" / "main.json"
        view_data = json.loads(view_file.read_text())
        assert view_data["context_files"] == []


def test_autocompletion_includes_symlinks(tmp_path: Path) -> None:
    # GIVEN session with symlinks and regular files
    with runner.isolated_filesystem(temp_dir=tmp_path) as td:
        runner.invoke(app, ["init"])

        # Regular file
        regular = Path(td) / "regular.py"
        regular.touch()

        # Symlinks
        sym1 = Path(td) / "sym1.py"
        sym1.symlink_to("regular.py")
        sym2 = Path(td) / "src/sym2.py"
        sym2.parent.mkdir()
        sym2.symlink_to("../regular.py")

        runner.invoke(app, ["add", "regular.py", "sym1.py", "src/sym2.py"])

        # WHEN completion
        # THEN includes symlinks (from session JSON paths)
        assert sorted(complete_files_in_context(None, [], "sym")) == ["sym1.py"]
        assert sorted(complete_files_in_context(None, [], "src/")) == ["src/sym2.py"]
        assert sorted(complete_files_in_context(None, [], "")) == sorted(["regular.py", "src/sym2.py", "sym1.py"])

    # GIVEN a directory with no session file
    with runner.isolated_filesystem():
        # WHEN the completion function is called
        completions = complete_files_in_context(None, [], "any")
        # THEN it returns an empty list without erroring
        assert completions == []

File: tests/test_command_context.rs
mod common;
use assert_cmd::cargo_bin_cmd;
use predicates::prelude::*;
use std::fs;
use tempfile;
use tempfile::tempdir;

#[test]
fn test_add_file_to_context() {
    let temp = tempfile::tempdir().unwrap();
    let root = temp.path();

    // Init
    let mut cmd = cargo_bin_cmd!("aico");
    cmd.current_dir(root).arg("init").assert().success();

    // Create file
    let file_path = root.join("test_file.py");
    fs::write(&file_path, "print('hello')").unwrap();

    // Add
    let mut cmd_add = cargo_bin_cmd!("aico");
    cmd_add
        .current_dir(root)
        .arg("add")
        .arg("test_file.py")
        .assert()
        .success()
        .stdout(predicate::str::contains(
            "Added file to context: test_file.py",
        ));

    // Verify View
    let view_path = root.join(".aico/sessions/main.json");
    let content = fs::read_to_string(view_path).unwrap();
    assert!(content.contains("test_file.py"));
}

#[test]
fn test_add_duplicate_file_is_ignored() {
    let temp = tempfile::tempdir().unwrap();
    let root = temp.path();

    cargo_bin_cmd!("aico")
        .current_dir(root)
        .arg("init")
        .assert()
        .success();
    fs::write(root.join("test.py"), "").unwrap();

    // Add once
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .args(["add", "test.py"])
        .assert()
        .success();

    // Add again
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .args(["add", "test.py"])
        .assert()
        .success()
        .stdout(predicate::str::contains("File already in context: test.py"));
}

#[test]
fn test_add_non_existent_file_fails() {
    let temp = tempfile::tempdir().unwrap();
    cargo_bin_cmd!("aico")
        .current_dir(&temp)
        .arg("init")
        .assert()
        .success();

    cargo_bin_cmd!("aico")
        .current_dir(&temp)
        .args(["add", "missing.py"])
        .assert()
        .failure()
        .stderr(predicate::str::contains("Error: File not found"));
}

#[test]
fn test_drop_single_file_successfully() {
    let temp = tempfile::tempdir().unwrap();
    let root = temp.path();

    cargo_bin_cmd!("aico")
        .current_dir(root)
        .arg("init")
        .assert()
        .success();
    fs::write(root.join("f1.py"), "").unwrap();
    fs::write(root.join("f2.py"), "").unwrap();

    cargo_bin_cmd!("aico")
        .current_dir(root)
        .args(["add", "f1.py", "f2.py"])
        .assert()
        .success();

    // Drop
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .args(["drop", "f1.py"])
        .assert()
        .success()
        .stdout(predicate::str::contains("Dropped file from context: f1.py"));

    // Verify f2 remains
    let content = fs::read_to_string(root.join(".aico/sessions/main.json")).unwrap();
    assert!(!content.contains("f1.py"));
    assert!(content.contains("f2.py"));
}

#[test]
fn test_add_multiple_files_successfully() {
    let temp = tempdir().unwrap();
    let root = temp.path();
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .arg("init")
        .assert()
        .success();

    fs::write(root.join("file1.py"), "").unwrap();
    fs::write(root.join("file2.py"), "").unwrap();

    cargo_bin_cmd!("aico")
        .current_dir(root)
        .args(["add", "file1.py", "file2.py"])
        .assert()
        .success()
        .stdout(predicate::str::contains("Added file to context: file1.py"))
        .stdout(predicate::str::contains("Added file to context: file2.py"));

    let content = fs::read_to_string(root.join(".aico/sessions/main.json")).unwrap();
    assert!(content.contains("file1.py"));
    assert!(content.contains("file2.py"));
}

#[test]
fn test_add_multiple_files_with_one_already_in_context() {
    let temp = tempdir().unwrap();
    let root = temp.path();
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .arg("init")
        .assert()
        .success();

    fs::write(root.join("file1.py"), "").unwrap();
    fs::write(root.join("file2.py"), "").unwrap();
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .args(["add", "file1.py"])
        .assert()
        .success();

    cargo_bin_cmd!("aico")
        .current_dir(root)
        .args(["add", "file1.py", "file2.py"])
        .assert()
        .success()
        .stdout(predicate::str::contains(
            "File already in context: file1.py",
        ))
        .stdout(predicate::str::contains("Added file to context: file2.py"));
}

#[test]
fn test_add_multiple_files_with_one_non_existent_partially_fails() {
    let temp = tempdir().unwrap();
    let root = temp.path();
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .arg("init")
        .assert()
        .success();

    fs::write(root.join("valid.py"), "").unwrap();

    cargo_bin_cmd!("aico")
        .current_dir(root)
        .args(["add", "valid.py", "missing.py"])
        .assert()
        .failure()
        .stdout(predicate::str::contains("Added file to context: valid.py"))
        .stderr(predicate::str::contains(
            "Error: File not found: missing.py",
        ));

    // Verify valid.py WAS still added despite the error
    let content = fs::read_to_string(root.join(".aico/sessions/main.json")).unwrap();
    assert!(content.contains("valid.py"));
}

#[test]
fn test_drop_multiple_files_successfully() {
    let temp = tempdir().unwrap();
    let root = temp.path();
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .arg("init")
        .assert()
        .success();

    fs::write(root.join("f1.py"), "").unwrap();
    fs::write(root.join("f2.py"), "").unwrap();
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .args(["add", "f1.py", "f2.py"])
        .assert()
        .success();

    cargo_bin_cmd!("aico")
        .current_dir(root)
        .args(["drop", "f1.py", "f2.py"])
        .assert()
        .success()
        .stdout(predicate::str::contains("Dropped file from context: f1.py"))
        .stdout(predicate::str::contains("Dropped file from context: f2.py"));
}

#[test]
fn test_drop_multiple_with_one_not_in_context_partially_fails() {
    let temp = tempdir().unwrap();
    let root = temp.path();
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .arg("init")
        .assert()
        .success();

    fs::write(root.join("f1.py"), "").unwrap();
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .args(["add", "f1.py"])
        .assert()
        .success();

    cargo_bin_cmd!("aico")
        .current_dir(root)
        .args(["drop", "f1.py", "missing.py"])
        .assert()
        .failure()
        .stdout(predicate::str::contains("Dropped file from context: f1.py"))
        .stderr(predicate::str::contains(
            "Error: File not in context: missing.py",
        ));
}

#[test]
fn test_add_symlink_to_inside_success() {
    let temp = tempdir().unwrap();
    let root = temp.path();
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .arg("init")
        .assert()
        .success();

    let target = root.join("target.py");
    fs::write(&target, "print(1)").unwrap();
    let link = root.join("link.py");
    std::os::unix::fs::symlink("target.py", &link).unwrap();

    cargo_bin_cmd!("aico")
        .current_dir(root)
        .args(["add", "link.py"])
        .assert()
        .success()
        .stdout(predicate::str::contains("Added file to context: link.py"));
}

#[test]
fn test_drop_symlink_success() {
    let temp = tempdir().unwrap();
    let root = temp.path();
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .arg("init")
        .assert()
        .success();

    fs::write(root.join("target.py"), "").unwrap();
    std::os::unix::fs::symlink("target.py", root.join("link.py")).unwrap();
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .args(["add", "link.py"])
        .assert()
        .success();

    cargo_bin_cmd!("aico")
        .current_dir(root)
        .args(["drop", "link.py"])
        .assert()
        .success()
        .stdout(predicate::str::contains(
            "Dropped file from context: link.py",
        ));
}

#[test]
fn test_drop_file_not_in_context_fails() {
    let temp = tempfile::tempdir().unwrap();
    let root = temp.path();
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .arg("init")
        .assert()
        .success();
    fs::write(root.join("f1.py"), "").unwrap();

    // Attempt to drop file not in context (file exists on disk)
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .args(["drop", "f1.py"])
        .assert()
        .failure()
        .stderr(predicate::str::contains(
            "Error: File not in context: f1.py",
        ));
}

#[test]
fn test_drop_file_missing_from_disk_but_in_context_success() {
    let temp = tempdir().unwrap();
    let root = temp.path();
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .arg("init")
        .assert()
        .success();

    let f = root.join("transient.py");
    fs::write(&f, "").unwrap();
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .args(["add", "transient.py"])
        .assert()
        .success();

    // Delete from disk
    fs::remove_file(&f).unwrap();

    // Drop should still work because it removes from metadata
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .args(["drop", "transient.py"])
        .assert()
        .success()
        .stdout(predicate::str::contains(
            "Dropped file from context: transient.py",
        ));

    let content = fs::read_to_string(root.join(".aico/sessions/main.json")).unwrap();
    assert!(!content.contains("transient.py"));
}

#[test]
fn test_drop_autocompletion() {
    let temp = tempdir().unwrap();
    let root = temp.path();

    // Init session
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .arg("init")
        .assert()
        .success();

    // Create and add files
    fs::write(root.join("file1.txt"), "content").unwrap();
    fs::write(root.join("file2.txt"), "content").unwrap();

    cargo_bin_cmd!("aico")
        .current_dir(root)
        .arg("add")
        .arg("file1.txt")
        .arg("file2.txt")
        .assert()
        .success();

    // Verify 'drop' suggests context files using clap's dynamic completion
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .env("COMPLETE", "bash")
        .env("_CLAP_COMPLETE_INDEX", "2")
        .args(["--", "aico", "drop", "file"])
        .assert()
        .stdout(predicate::str::contains("file1.txt"))
        .stdout(predicate::str::contains("file2.txt"));
}

#[test]
fn test_autocompletion_includes_symlinks() {
    let temp = tempdir().unwrap();
    let root = temp.path();

    cargo_bin_cmd!("aico")
        .current_dir(root)
        .arg("init")
        .assert()
        .success();

    let target = root.join("real_file.txt");
    fs::write(&target, "data").unwrap();

    #[cfg(unix)]
    {
        let link = root.join("link_file.txt");
        std::os::unix::fs::symlink("real_file.txt", &link).unwrap();

        cargo_bin_cmd!("aico")
            .current_dir(root)
            .arg("add")
            .arg("link_file.txt")
            .assert()
            .success();

        let session = aico::session::Session::load(root.join(".ai_session.json")).unwrap();
        let context = session.get_context_files();

        // Must preserve the link name, not resolve to target
        assert!(context.contains(&"link_file.txt".to_string()));
        assert!(!context.contains(&"real_file.txt".to_string()));
    }
}

#[test]
fn test_drop_is_silent_on_missing_files() {
    let temp = tempdir().unwrap();
    let root = temp.path();

    cargo_bin_cmd!("aico")
        .current_dir(root)
        .args(["init", "--model", "test-model"])
        .assert()
        .success();

    // GIVEN a file tagged in context but missing from disk
    let view_path = root.join(".aico/sessions/main.json");
    let mut view: aico::models::SessionView =
        serde_json::from_str(&fs::read_to_string(&view_path).unwrap()).unwrap();
    view.context_files.push("missing.py".to_string());
    fs::write(&view_path, serde_json::to_string(&view).unwrap()).unwrap();

    // WHEN dropping it
    let assert = cargo_bin_cmd!("aico")
        .current_dir(root)
        .args(["drop", "missing.py"])
        .assert()
        .success();

    // THEN it should NOT emit the "not found on disk" warning to stderr
    assert
        .stderr(predicate::str::contains("Warning: Context files not found on disk").not())
        .stdout(predicate::str::contains(
            "Dropped file from context: missing.py",
        ));
}
